### DISCUSSION TITLE
Numerical habits are not cumulative and require the goal to be completed in one day

### DISCUSSION THREAD

**Discussion by habitswidget** [iSoron/uhabits#1180] [General Discussion]
## Pre-submission checklist
- [y] I am submitting a bug report, not a feature request.
- [y] I have searched for similar issues, but did not find any matches. 

There is another bug report related to measurable habits and the Hide Completed function - I have Hide Completed unchecked throughout.

## Description
I created a habit to run 6km per week and completed 3km on two separate days this week. Widget doesn't mark as complete. But it will mark as complete if I ran 6 today.

## Steps to reproduce
1. Create new measurable habit with a target of 2 or more per week.
2. Create Widget for Habit
3. Mark as complete on 2 days within the week where the sum is equal to or greater than the goal (but neither are on their own)
4. See that the widget has not marked as completed

## System information
 - Phone: Pixel 3a
 - Phone Operating System: Android 11
 - App version: 2.0.3

## Screenshots

<img width="291" alt="Screenshot 2021-10-07 at 13 50 48" src="https://user-images.githubusercontent.com/92097138/136387511-d64e84b9-ca07-43c8-962d-2e969889cd33.png">
<img width="389" alt="Screenshot 2021-10-07 at 13 50 15" src="https://user-images.githubusercontent.com/92097138/136387526-d6713eb9-5513-4164-bf9b-eb8809f10423.png">

Thanks for your help, love the app.

**Comment by hiqua** [comment]
I'm not sure how it was intended, but I think your expectation of being able to reach your goal over several days is common.

It could be that we just have to add one more type of habits, or replace the current one by a cumulative numerical habit instead.

@iSoron what was your initial intent? Should we have a new type of habit, or just change the current behavior to make it cumulative?

To compare with yes/no habits: if you did 10 when your target is 20, you wouldn't check a habit asking "did you do 20 today?". So in a way, the current implementation of numerical habits match these more accurately than a new hypothetical cumulative numerical habit. I probably would prefer a cumulative behavior myself though.

**Reply by iSoron** [reply]
@hiqua We should make it cumulative. Instead of just checking today's value, we should sum over the previous days and compare against the target.

**Reply by VA-MMXXII** [reply]
It may be that a hybrid type is called for here, or new options made available to the numerical type such that if x units is achieved goal is met and a checkmark is added to the numerical value.

**Comment by ke352802081770314** [comment]
I must admit I am puzzled by the report and discussion because I am convinced that scoring already works cumulatively.
I have a habit "Gardening" with goal 150 (minutes) every week, and I have been scoring quite good on it (over 90 percent most of the time, deteriorating a bit now in winter) but I rarely did 150 min on a single day. So to me it feels like the scoring mechanism looks at the sum of each day's number for a weekly period. This also matches the progress bars for week,month,quarter,year (except for the fact that these are not sliding-window but based on the calendar and therefore now at the beginning of the new year look quite poor).
However, unlike on/off habits, there is neither the percentage score shown on the habit view, nor the streaks date periods, the calendar seems to have full color for each day where anything more than zero was entered, so it is hard to tell how it was evaluated.
I do understand that the coloring of each day's number on the overview follows some rule and works on a per-day basis without looking at the reaching of the goal, there is no equivalent to an outline checkmark because the space is used for showing the number, which seems logical to me.

So maybe it is not an issue with scoring but only with a lack of feedback shown?
Disclaimer: I still run 2.0.2, sorry about that.

**Reply by jamesdeluk** [reply]
Are you talking about the overall progress screen or the daily value?

As I see it, if that week's target has been met, even the 0s (or no entry ones) should change colour, to show "success". This does not currently happen (unlike with the YES/NO where if the target is met the Xs or no entries become outline ticks).

For the progress stats, that all depends on the time. I feel there needs to be a function similar to "start measuring progress from 2022/02/01" or similar, as otherwise it always takes a long time to get a high score as it climbs from 0 to 100 instead of starting at 100 and dropping to 0 on failures. But that's a different issue that may or may not be discussed elsewhere :)

**Reply by ke352802081770314** [reply]
you are right that in the list, the numbers are only coloured on an individual-day basis with light grey for ?/zero, dark gray up to some obscurely calculated number, and in the goal's colour above that. The "outline checkmark" has no correspondence here, and I agree that the colouring should be adjusted taking into account the goal's fulfilment period (which, for numerical goal, only offers day/week/month whereas with yes/no goals I have the option "every N days".

I consider this cosmetic because the score, shown as the partial circle to the left of the goal name, and in the "scoring" time graph (I hope it is called scoring, it's "Wertung" in German) below the week/month/quarter/year progress bars on the individual goal's page, appears to be calculated as I expect. I also find it unfortunate that the first section for yes/no goals, "Overview" with the circle, current score, difference month and year, total is missing from the goal view for numerical goals. it should be present for them just the same, but I think we already have a separate ticket asking for that.

On your more fundatental point, in essence, "start measuring progress" already happens automatically from the first data point you enter. 
A habit is formed by regularly performing it. There is no credit for "I would have been perfect I just didn't aim for it at all", and rightly so, there shouldn't be a default perfect history, that I can gain just by entering a start date. Only actual repetitions will ingrain the habit into your body (or subconscious or whereever it is located) and make it a ritual that is at some point performed "automatically". If you start from scratch, building a habit and sticking to it until it becomes "automatic" is a tedious and long process, this is simply a fact. Loop Habit Tracker reflects this fact in its scoring mechanism.
I think this is more rewarding than the other way round. If by entering a start date you would gain a perfect score to begin with, every lapse would be a disappointment, and lapses are to be expected until a habit has been established because it hasn't become routine yet. If you manage to stick to it consistenly, you will see your score increasing all the time until it eventually reaches 100% (the last 10% are the hardest, this is just how life is, you can't be sure you are perfect until you've performed very well for quite a while). Imagine on the other hand being given 100% credit in advance and deteriorating to 70% due to "a few" slip-ups. Wouldn't that be daunting and you might think "to hell with it, everything I touch goes awfully wrong"? 
In case you already started and were on a good way before starting to use habit tracker, or entering that particular goal, feel free to enter your history into past days and you will be rewarded with the score "you deserve". ;)
How would you justify a start value of 100 and what benefits do you see in that, that would outweigh the disadvantages I hope I could illustrate? why not just enter a starting date next year and be given a perfect score this year as a reward? isn't that what procrastination is all about?
(sorry for discussing it here anyway, you didn't provide a pointer to the more appropriate place.)

**Reply by jamesdeluk** [reply]
You make a very good point. I use the app more for general tracking rather than purely habits, where it's not about building a habit directly (which requires the long-term commitment as you described), but simply knowing how many times I have reached my target since starting. I don't use the progress section particularly often anyway, and when I do I look at the bottom graph (History), so I'm not too concerned over the 0-100 aspect.

However, changing the colour of the ?/0 and similar records to a "success" colour would be very helpful for my use - cosmetic, yes, but very helpful to know if I've hit my goal for that week or not, and hence if I need to prioritise it.
In other words, if my goal is run 20km a week, blue in colour, and Monday I do 5, Tuesday 4, Wednesday I do 1, and Friday 10, Thursday, Saturday, and Sunday should all be ?/0 and blue in colour (as in the week I have hit my goal of 20km) - currently they are not.
![Screenshot_20220206-074626](https://user-images.githubusercontent.com/27723508/152672026-c7aa739d-518e-46e5-9491-9570429aa94f.jpg)

**Reply by ke352802081770314** [reply]
you are right, visual feedback on the main page is valuable, and therefore my calling it cosmetic was not only subjective but a wrong judgement. I agree that would be beneficial.

**Comment by baincd** [comment]
What I'd really like (and I think others in this discussion as well) is basically the ability to track a habit like a Yes/No habit, but using a numerical value rather than just Yes/No.  For now, I'm going to call this a Summation habit.

An example of a Summation habit would be to do at least 150 minutes of aerobic exercise every 7 days.  This Summation habit would:
- Track number of minutes exercising each day
- For each day, add up the number of minutes of exercise for that day plus the previous 6 days.  If that number is greater than or equal to 150, consider the habit as "completed" for that day (just like the open-checkmark on Yes/No habits) - even if a 0 was tracked for that day
- The app determines streaks and calculates the habit score the same as Yes/No habits (using the logic above to determine days where the habit was "completed")
- I don't really see the need for the Target information that Measurable habits have, as I'm more interested in streaks.  But others may feel differently.

In fact, Yes/No habits could be thought of as a specialized case of a Summation habit, just where Yes=1 and No=0.

I don't know if Measurable habits could be modified to support Summation habits as well, since Measurable habits seem to be pretty different.  Summation habits may need to be a third habit type, or an option on a Yes/No habit.

(FYI, I am not attached to the name Summation habit - I just needed to name it something to make it easier to reference)

**Comment by amrsamii** [comment]
Any update on this? The problem still exists.

**Comment by KyleSCraig** [comment]
I've done some work on this, but I think the direction I've gone down is the wrong way to address it. I got the Calendar view within a habit working with both (if you do more than target/frequency it's solid) and (if you have done the target within the frequency period it's dimmed). For the front page I could just get (if you do more than target/frequency it's solid) and I don't see an easy way to do what I did in the calendar. I also had a brief look at streaks and I think that's also more complex than Calendar.

Instead of just trying to address it at the visual level, I believe something like YES_AUTO needs to be implemented for numerical habits in the backend.

**Reply by KyleSCraig** [reply]
I have implemented more of it, and have indeed implemented it like YES_AUTO. It's meeting my needs, but still has issues.

But I have realised that I have no chance of getting it merged because I have worked on it to be similar to Yes No habits and completely disregarded the whole "Target" system. Personally I generally prefer to have rolling windows than blocks like the target. I kind of think the direction it needs to go in is to allow rolling and block periods for both habit types, but that's a lot of work.

**Comment by oblqn** [comment]
Hello there! I am also very interested in this feature.

On top of everything that has been said, I would argue that this "cumulative sum" painting scheme is already what is being done for binary habits. For such habits, the painting scheme of a given day on the calendar seems to be:
     if (data_is_entered) -> strong color
     elseif (count_in_window >= habit_target) -> light color
     else -> no color
Where count_in_window reads as "number of days that the habit has been done in the preceeding habit window" (the window is of course the time lapse specified by the user when defining the habit frequency).

It would look to me like the natural extension of this painting scheme from the binary to the measurable habits is to do the same counting, now adding X instead of 1 for each entry, with X the entered habit value.

**Comment by keshavnarla** [comment]
Hello,

I would love to see this feature. I have goals like 10 pull-ups per week and marking 3/day and skip Saturday, Sunday. 

So my total per week is more than 10 but there is no streak showing up unless I mark 10 pullupd in a day

Is there a workaround for this?

Thanks!

**Comment by mintyvoid** [comment]
Is there any plan to change this? It's odd that the measurable habits doesn't function like you would think it would at least visually still. The checkmark widget (and the circle in the app itself) doesn't seem to fill based on the target and that's literally all I want.
